src/core/shader.js

```diff
@@ -56,7 +58,8 @@ Shader.prototype = {
       // attributes: this.attributes,
       uniforms: this.uniforms,
       vertexShader: this.vertexShader,
-      fragmentShader: this.fragmentShader
+      fragmentShader: this.fragmentShader,
+      glslVersion: this.glslVersion
```

Comment:
probably passing here `THREE.GLSL3` is all the changes we need if I'm not missing anything. thoughts?

Reply:
Hardcoding version instead of making default value for parameter can be not the best solution. What if somewhere in future we will get GLSL4,5,6 or sort of custom extensions which require to pass different version string? For me the right way was is respect original user shader code but THREE did this...

Reply:
Hardcoding in this instance is not too bad. There's almost certainly never going to be a newer version for WebGL with WebGPU on the horizon.

It all boils down to two separate decisions (see other comment thread): do we want to support ES 1.0 GLSL shaders with A-Frame and do we want to support WebGL1 (if that even works right now?).

Reply:
We can hardcode for now or can also determine from the renderer (WebGL1, WebGL2, WebGPU) instead of a public API. sounds good? thanks for the patience

Reply:
Did some testing and it seems A-Frame can (still) work fine with WebGL1, so I'd rather not break core components and try to postpone that till Three.js drops support (r163). Also found out that hardcoding it, actually breaks non-raw custom shaders as the GLSL 3.0 shader conversion that Three.js does ends up behaving slightly different :-/

Automatically determining it from the renderer won't be an option as the shader _is_ written in a specific shading language. So either we expose a way for users to tell which language/version they used, or we restrict what A-Frame supports.

Basically the following support matrix exists:
|         | WebGL 1.0 | WebGL 2.0 | WebGPU |
|-------|------------------|-----------------|--------------|
|ShaderMaterial (GLSL 1.0 /w 3.0 conversion) | ✅/❌ (2)| ✅ | ❌ |
| GLSL 1.0 | ✅ | ✅ | ❌ |
| GLSL 3.0 | ❌ | ⚠️ (1) | ❌ |
| NodeMaterial | ? | ✅ | ✅ |

 1. Before the Three.js update the shader code itself could indicate the GLSL version in its source. Now that three.js injects other things at the start, it _must_ also handle injecting the version directive. By extension this means that we _must_ tell three.js which version it is.
 2. Depends on which GLSL features and keywords are used. The GLSL 3.0 conversion in Three.js only handles a small subset.

### **tl;dr** I'd suggest the following:

- For now set `glslVersion: raw ? THREE.GLSL3 : undefined`, fixing the problem for the majority of the users.
- As a separate/follow-up: change the sdf and msdf shaders to non-raw shaders. This would remove the need to keep two almost identical copies around (GLSL1 and GLSL3 versions), restores WebGL1 compatibility, makes it easier to fix fog interaction (#4261), logarithmic depth (#4279) and color management.

Reply:
let's go for the simple solution. thanks

> For now set glslVersion: raw ? THREE.GLSL3 : undefined, fixing the problem for the majority of the users.
> As a separate/follow-up: change the sdf and msdf shaders to non-raw shaders. This would remove the need to keep two almost identical copies around (GLSL1 and GLSL3 versions), restores WebGL1 compatibility, makes it easier to fix fog interaction (https://github.com/aframevr/aframe/issues/4261), logarithmic depth (https://github.com/aframevr/aframe/issues/4279) and color management.